### PR TITLE
interfaces: share /tmp/.X11-unix/ from host or provider

### PR DESCRIPTION
**Pull request by zyga**
This relinquishes the reliance on the abstract sockets for X11
connectivity, fixing X11 snaps under Wayland on Groovy.

Signed-off-by: Zygmunt Krynicki <me@zygoon.pl>

### CODE REVIEW

**Comment by zyga** [comment]
@alexmurray we'd like to bring the host's /tmp/.X11-unix directory, along with the sockets there, over to the snap world. This will allow X11 snaps to work on systems without the abstract unix socket. Could you please do the security assessment of such change?

File: interfaces/builtin/x11.go
Comment (jhenstridge) [inline interfaces/builtin/x11.go:157]: This is clearly correct for the `implicitSystemPermanentSlot(slot)` case, but not otherwise.  In that case, the source probably needs to be `/var/lib/snapd/hostfs/tmp/snap.${slot_snap}/tmp/.X11-unix`.  That probably requires a `MountConnectedPlug` implementaqtion rather than the default `commonInterface` code.  Ideally this would use some kind of helper to determine the slot snap's private tmp directory path, to avoid getting out of sync with snap-confine.

I think it would be worth making this a read-only bind mount too, to prevent the plug snap from modifying the directory contents.

File: interfaces/builtin/x11.go
Comment (jhenstridge) [inline interfaces/builtin/x11.go:166]: As above, this will need to be slightly different for non-implicit `x11` slots.

**Review by jhenstridge — CHANGES_REQUESTED** [review]
Looks like a good start.

Would it make sense to include some AppArmor protections for the `/tmp/.X11-unix` directory on the plug side?  Currently we have `#include <abstractions/X>`, which grants read/write to files within that directory.  I believe that's necessary for socket communication, but it would be good to prevent moving or deleting the directory itself, since the snap otherwise has complete control over `/tmp`.

**Comment by jhenstridge** [comment]
I was mistaken about being able to move the destination of the bind mount: it's the source that can be moved.  Other than breaking themselves, I'm not sure there's much nefarious that the slot snap could do with that other than breaking themselves.

One thing that would definitely need to be checked: if the slot snap's private temporary directory does not exist when building the plug snap's mount namespace, what happens?  I understand the default is to create the missing directories, but that leaves the following open questions:

1. will the `/tmp/snap.${slot_snap}` and `/tmp/snap.${slot_snap}/tmp` directories be created with permissions preventing non-root users on the host system from getting up to mischief?
2. will the `/tmp/.X11-unix` directory in the slot's sandbox be writable by non-root users?

**Comment by jhenstridge** [comment]
Another case to test: the x11 plug and slot are on the same snap.  While bind mounting a directory over itself is not a problem in itself, if we make it a read-only bind mount the X server won't be able to create its sockets.

File: interfaces/builtin/x11.go
Comment (Erick555) [reply interfaces/builtin/x11.go:157]: yes mount read-only is much appreciated here.

Side note: `abstractions/X` is currently broken in AppArmor 3.0+ because socket doesn't [have write access](abstractions/X) which is surprisingly needed in apparmor semantics.

File: interfaces/builtin/x11.go
Comment (jhenstridge) [inline interfaces/builtin/x11.go:161]: Could you use implicitSystemConnectedSlot(slot) here?  That would avoid adding more ways of detecting implicit slots.

**Review by jhenstridge — COMMENTED** [review]
Looks better.  I expect you'd still need some extra code in snap-update-ns to match permissions like this, if it ends up creating the directory tree:

```
drwx------ 3 root root 4096 Oct  5 13:46 /tmp/snap.chromium
drwxrwxrwt 3 root root 4096 Oct  5 13:46 /tmp/snap.chromium/tmp
```

**Comment by zyga** [comment]
@jhenstridge The permissions on various components of the path in /tmp are likely to be problematic. I need to think about our options.

**Comment by zyga** [comment]
Still largely untested but what I just pushed should use the right ownership of the created `/tmp/snap.*` and `/tmp/snap.*/tmp`

File: interfaces/builtin/x11.go
Comment (zyga) [reply interfaces/builtin/x11.go:161]: Done

File: interfaces/builtin/x11.go
Comment (zyga) [reply interfaces/builtin/x11.go]: I think this is done.

CC @alexmurray and @jdstrand about apparmor X abstraction and `w+` permission on sockets? Will the read-only bind mount pose a problem here?

File: interfaces/builtin/x11.go
Comment (zyga) [reply interfaces/builtin/x11.go]: Done

**Comment by zyga** [comment]
> Looks better. I expect you'd still need some extra code in snap-update-ns to match permissions like this, if it ends up creating the directory tree:
> 
> ```
> drwx------ 3 root root 4096 Oct  5 13:46 /tmp/snap.chromium
> drwxrwxrwt 3 root root 4096 Oct  5 13:46 /tmp/snap.chromium/tmp
> ```

This is now done.

File: interfaces/builtin/x11.go
Comment (Erick555) [reply interfaces/builtin/x11.go]: read-only mount should work, flatpak does it and other sandbox frameworks too.

**Comment by zyga** [comment]
I'll park this until next week. This needs a security review now. Following that we need a few more tests (basic unit tests and integration test, primarily for the slot-provided-x11 case.

File: cmd/snap-update-ns/system.go
Comment (jhenstridge) [inline cmd/snap-update-ns/system.go:83]: Should there be a hint for `/var/lib/snapd/hostfs/tmp/snap.*/tmp/.X11-unix` too?  We'd want it created with permissions allowing normal users to create sockets in it.  It should have `1777` permissions too.

**Review by jhenstridge — COMMENTED** [review]
Looks good.  It would be good to have some spread tests to verify the following:

1. That a plug snap cannot modify the contents of `/tmp/.X11-unix` directory.
2. That a slot snap can create files in `/tmp/.X11-unix` as a regular user if the directory was created during setup of a plug snap's mount namespace.

File: cmd/snap-update-ns/system.go
Comment (zyga) [reply cmd/snap-update-ns/system.go:83]: Hmm, are you saying that the slot provider will non be a root process? If so then yes, I will amend this section.

Comment (zyga) [reply cmd/snap-update-ns/system.go:83]: Done

**Comment by zyga** [comment]
I agree about tests. I will do all the tests next week as I need to switch gears to another task today.

File: cmd/snap-update-ns/system.go
Comment (jhenstridge) [reply cmd/snap-update-ns/system.go:83]: Definitely.  X servers are usually run as a regular user on modern Linux systems.  If its Xorg running on direct hardware, it generally gains access to the hardware via logind (which we can grant access to with `login-session-control`).  If it's Xwayland, there's no special hardware access needed, so again easy to run as a normal user.

**Comment by Erick555** [comment]
@zyga are you sure about ICE stuff? flatpak doesn't mount that and I didn't saw any complaints from users.

**Comment by jhenstridge** [comment]
I would suggest leaving out the ICE-unix sockets too.  There is a good argument that we shouldn't be granting access to the abstract version either.  I doubt many snaps make use of XSMP, and it likely isn't secure.

**Review by jhenstridge — CHANGES_REQUESTED** [review]
I think the `/tmp/.ICE-unix` part of this PR should be dropped (or at least discussed).  Otherwise, it looks pretty good.  Do you want to handle spread tests in this PR, or as a follow-up?

**Comment by zyga** [comment]
I added ICE because it is a part of the interface. Perhaps the interface should be revised?

**Comment by zyga** [comment]
I will do spread tests here, most likely later today.

**Comment by alexmurray** [comment]
From a security point-of-view, ICE exposes the session management APIs amongst others which could be used to escape the snap sandbox (by instructing the session manager to launch the snap outside of confinement etc). This should be removed unless there is a specific compelling use-case for it to be included. By just including `/tmp/.X11-unix` in this interface this also keeps it to parity with the access that is granted by the abstract socket as well.

**Comment by zyga** [comment]
I've removed parts of ICE from this pull request. I will send separate ICE-specific changes to the base of the x11 interface in a separate pull request.

**Comment by zyga** [comment]
I'm working on testing this now.

**Comment by ilja903** [comment]
Sad to see that almost clean wayland ubuntu 20.04 became unusable after upgrade to 20.10. Snap is a great idea, but the implementation could be tested more carefully. Idk, it is just sad that even chromium would not start.

**Comment by jhenstridge** [comment]
@ilja903: note that we've also been working on this problem from the GNOME side in [mutter MR !1508](https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1508).  That change has been accepted and will be in a future 3.38.x release.  We'll probably have the fix released as an update for Ubuntu 20.10 before then though.  I agree that it would have been nice if we'd diagnosed this earlier.

This PR is more about the long term, since upstream has indicated that they'd like to make the abstract socket removal permanent in a future GNOME release.

**Comment by zyga** [comment]
And we need some fixes for SELinux as well.
```
type=SYSCALL msg=audit(10/27/20 12:31:35.999:23751) : arch=x86_64 syscall=unlinkat success=yes exit=0 a0=0xffffffffffffff9c a1=0xc0000fba80 a2=0x200 a3=0xc0000fba80 items=0 ppid=9268 pid=12800 auid=unset uid=root gid=root euid=root suid=root fsuid=root egid=root sgid=root fsgid=root tty=(none) ses=unset comm=snap-update-ns exe=/usr/libexec/snapd/snap-update-ns subj=system_u:system_r:snappy_mount_t:s0 key=(null) 
type=AVC msg=audit(10/27/20 12:31:35.999:23751) : avc:  denied  { rmdir } for  pid=12800 comm=snap-update-ns name=.X11-unix dev="sda2" ino=17778788 scontext=system_u:system_r:snappy_mount_t:s0 tcontext=system_u:object_r:tmp_t:s0 tclass=dir permissive=1 
```

File: interfaces/builtin/x11.go
Comment (Erick555) [reply interfaces/builtin/x11.go]: abstractions/X will be [fixed in next apparmor 3 release](https://gitlab.com/apparmor/apparmor/-/commit/0cb35fda84a6ace742d9da3a7630a0dcc6ffae9d)

**Comment by zyga** [comment]
I really give up on the centos 7 denial. @bboozzoo if you have a moment please try fixing this. I'm lost.

**Comment by AlanGriffiths** [comment]
I'm flagging my interest in this: various Mir based snaps contain workarounds for the invisibility of `/tmp/.X11-unix/` when choosing a DISPLAY for use by the snap. I have to understand the impact of this change on these workarounds.

For example: https://github.com/MirServer/mir-kiosk-x11-snap/blob/main/env-setup/hacks/setup-x11 has:
```plain
# We need to stop Mir accessing any existing X11 session.
# Mir detects existing sessions by scanning /tmp/.X11-unix
# but we don't have the real /tmp, but a snap specific one.
# So fake the existing $DISPLAY in the fake /tmp/.X11-unix
if [ -z "${DISPLAY}" ]
then touch /tmp/.X11-unix/X${DISPLAY:1}
fi
# A few more for luck
touch /tmp/.X11-unix/X0
touch /tmp/.X11-unix/X1
touch /tmp/.X11-unix/X2
touch /tmp/.X11-unix/X3
touch /tmp/.X11-unix/X4
```

**Comment by AlanGriffiths** [comment]
This breaks the following usecase used by a number of snaps that do both of the following:

1. Runs an Xorg server (Xwayland) internally (using the network-bind interface) with a non-abstract socket created in the private `/tmp/.X11-unix`
2. Plug the X11 interface to be able to use the session DISPLAY via the abstract socket

With these changes it is no longer possible to create the non-abstract socket as `/tmp/.X11-unix` is no longer the private one. **I realize that is rather the point.**

Running an "internal" Xorg server is the recommended approach for making "kiosk snaps" from X11 applications, there are a number of snaps using this approach.

There are fewer that also plug the X11 interface (which activates these changes): egmde-confined-desktop and mir-test-tools are two I know of.

A simple test case:
```plain
snap install mir-kiosk && snap install mir-test-tools --edge
snap connect mir-kiosk:x11
snap connect mir-test-tools:x11
WAYLAND_DISPLAY=wayland-99 mir-kiosk&
MIR_SERVER_WAYLAND_HOST=wayland-99 mir-test-tools.performance-test --gtest_filter=GLMark2Xwayland*
```
Disconnecting `mir-test-tools:x11` allows this case to work, but breaks the simpler case of running on the session DISPLAY:
```plain
mir-test-tools.performance-test --gtest_filter=GLMark2Xwayland*
```

**Comment by Erick555** [comment]
Would if work if X11 interface made binding `/tmp/.X11-unix` optional, i.e. if `/tmp/.X11-unix` already exist in snap then it would do nothing?

**Review by AlanGriffiths — CHANGES_REQUESTED** [review]
Rather than plugs mounting the plug `/tmp/.X11-unix`, would it be possible to instead link the entries in it? This would allow snaps that also create their own Xorg servers to create additional entries.

(This is the approach taken with snaps using the wayland interface - although not implemented in the interface itself.)

**Comment by AlanGriffiths** [comment]
> Would it work for you if X11 interface made binding `/tmp/.X11-unix` optional, i.e. if `/tmp/.X11-unix` already exist in snap then it would do nothing?

Not with a desktop that doesn't support abstract sockets

**Comment by Erick555** [comment]
But you said that you create this socket on your own so you don't need host one (and the conflict between those is the problem for you). If snap mount will be optional then you could keep doing what you did for the time being and make your app transition to new x11 interface behavior when you are ready (if you want).

**Comment by AlanGriffiths** [comment]
> But you said that you create this socket on your own so you don't need host one (and the conflict between those is the problem for you).

I said that I want to connect to the session `DISPLAY=:0`, which the GNOME changes worked via the abstract socket `@/tmp/.X11-unix/X0`.

The point of this PR is to accommodate desktop environments that do not support abstract sockets and so makes the `/tmp/.X11-unix/X0` non-abstract socket available. The scenario I describe will need this socket on these desktops.

The issue I'm describing is that by bringing the whole `/tmp/.X11-unix/` directory into the snap namespace the current proposal prevents the creation of another `DISPLAY=:5` non-abstract socket (`/tmp/.X11-unix/X5`) in that directory. Something a number of snaps rely on.

The solution I've suggested above is to only bring the `/tmp/.X11-unix/X0` socket into the snap namespace. (Hard coding this path into the interface locally suggests this would work, but we may want to support other sockets.)

**Comment by Erick555** [comment]
So what I proposed in https://github.com/snapcore/snapd/pull/9530#issuecomment-717951331 should work for you. Instead of hardcoding `/tmp/.X11-unix/X0` I think it should check actual socket name/address and mount it.

**Comment by AlanGriffiths** [comment]
> So what I proposed in [#9530 (comment)](https://github.com/snapcore/snapd/pull/9530#issuecomment-717951331) should work for you. Instead of hardcoding `/tmp/.X11-unix/X0` I think it should check actual socket name/address and mount it.

Yes, in fact my hacked version is bind mounting the socket, not linking. It should, of course, be checking the name(s) for production code.

**Comment by zyga** [comment]
@AlanGriffiths I'm not sure bind-mounting lock files will work reliably. Depending on the exact locking semantics, this may have the opposite result. As to bind-mounting just the X0 socket, that is possible but I think this may have unintended consequences for compatibility on the desktop.

Do you think it would be sensible to have an attribute on the x11 plug, which defines the behavior a particular snap is after?

**Comment by AlanGriffiths** [comment]
> @AlanGriffiths I'm not sure bind-mounting lock files will work reliably. Depending on the exact locking semantics, this may have the opposite result. 

I agree that could be tricky. And could be argued to be pre-existing.

The reason I mention it is that Xorg's os/utils.c has logic to remove sockets without the corresponding lockfile. If anyone ever uses an Xorg based server to serve X11 within a snap this could lead to unintended consequences. (This a hypothetical problem, it isn't a problem with the snaps I know of as Mir avoids this behavior, but it is something to be conscious of when weighing options.)

> As to bind-mounting just the X0 socket, that is possible 

Nitpicking: it isn't always `X0`, it's `X${DISPLAY:1}` where `DISPLAY` comes from the user desktop session of plugging snap.

Also it is probably a bad idea not to make `X*` visible, if not accessible (i.e. all X11 servers on the system). There's potential for confusion if there are two or more desktops sessions with abstract sockets running on, say, `X0` and `X1` and a snap connecting to `X0` chooses `X1` to create a local non-abstract socket. X11 clients in that snap could try to connect to the other desktop's abstract socket.

> but I think this may have unintended consequences for compatibility on the desktop.

Do you have a scenario in mind?

> Do you think it would be sensible to have an attribute on the x11 plug, which defines the behavior a particular snap is after?

I'd prefer to avoid that. Options lead to additional testing and a larger range of potentially problematic scenarios.

With this proposal as it stands we already have different behavior depending on whether the X11 interface is connected.

**Comment by zyga** [comment]
@AlanGriffiths (sorry for the format, I cannot hit reply on your comment somehow)
You said: Do you have a scenario in mind?

I was worried that bind-mounting a specific socket is racy and may not work reliably. When we bind mount the whole `.X11-unix` directory, in this specific case here, we do it in a way that allows the client to start, construct its mount namespace (and perhaps quit) in a way that is not broken. The `.X11-unix` directory for both the client and the server will be the same directory, same inode. I don't think this property would hold if we were sharing just `.X11-unix/X0`, because unless that is already a socket, we would create and share a placeholder file. What is worse, neither side would be able to remove that file, as it is a mount point.

**Comment by zyga** [comment]
@AlanGriffiths separately, I think we should also change interface permissions, as currently non-abstract sockets are not allowed. Perhaps I'm not reading this right or the permissions were changed in groovy (I've switched to focal now), but at least in focal only the abstract namespace socket is permitted. It's likely nobody noticed while testing on Wayland-SELinux system.

**Comment by AlanGriffiths** [comment]
> @AlanGriffiths separately, I think we should also change interface permissions, as currently non-abstract sockets are not allowed.

You mean clients in the plugging snap cannot connect to the non-abstract socket? Currently, they can't see the file in the snap specific `/tmp/.X11-unix`, so permissions are only relevant with these changes.

**Comment by zyga** [comment]
> > @AlanGriffiths separately, I think we should also change interface permissions, as currently non-abstract sockets are not allowed.
> 
> You mean clients in the plugging snap cannot connect to the non-abstract socket? Currently, they can't see the file in the snap specific `/tmp/.X11-unix`, so permissions are only relevant with these changes.

Yes, I agree.

**Comment by Erick555** [comment]
> I was worried that bind-mounting a specific socket is racy and may not work reliably.

@zyga As I wrote this is [what flatpak does ](https://github.com/flatpak/flatpak/blob/master/common/flatpak-run.c#L208) and I didn't heard of any issues coming from there.

**Comment by AlanGriffiths** [comment]
> I was worried that bind-mounting a specific socket is racy and may not work reliably. When we bind mount the whole `.X11-unix` directory, in this specific case here, we do it in a way that allows the client to start, construct its mount namespace (and perhaps quit) in a way that is not broken. The `.X11-unix` directory for both the client and the server will be the same directory, same inode. I don't think this property would hold if we were sharing just `.X11-unix/X0`, because unless that is already a socket, we would create and share a placeholder file. What is worse, neither side would be able to remove that file, as it is a mount point.

Yes, this is messier than the Wayland case where link files (created in `desktop_launch` or similar) are a working solution.

One option is to monitor the real `.X11-unix` directory for file creation and create bind mounts as that happens. And yes, that's potentially racy. (OTOH in practice it is unlikely for a snap to be launched in a client session before the X11 socket is established.)

**Comment by zyga** [comment]
> > I was worried that bind-mounting a specific socket is racy and may not work reliably.
> 
> @zyga As I wrote this is [what flatpak does ](https://github.com/flatpak/flatpak/blob/master/common/flatpak-run.c#L208) and I didn't heard of any issues coming from there.

Flatpack does not implement persistence of mount namespaces. We do and I can tell you there is a way to observe broken behaviour.

**Comment by zyga** [comment]
> One option is to monitor the real `.X11-unix` directory for file creation and create bind mounts as that happens. And yes, that's potentially racy. (OTOH in practice it is unlikely for a snap to be launched in a desktop session before the X11 socket is established.)

We have no facility today to do that.

**Comment by zyga** [comment]
>  (OTOH in practice it is unlikely for a snap to be launched in a desktop session before the X11 socket is established.)

It is sufficient for a snap to refresh and have a configure hook before a user logs in.

**Comment by AlanGriffiths** [comment]
@zyga as this stands it breaks stuff that happens to work right now. So breaking it should be intentional.

Also, @Erick555 has a good point in seeking consistency with Flatpak, DEs and apps will expect the same things to work.

But, ultimately, the stuff that breaks isn't of vital importance to any usage I know of. And I see the issues you raise with the alternative approach I suggested.

**Comment by bboozzoo** [comment]
Please use squash merge when landing the PR to make the cherry-picking to distro packages easier.

**Comment by zyga** [comment]
I've asked @bboozzoo for help with the remaining SELinux issue.

**Comment by zyga** [comment]
@anonymouse64 can you please describe the exact blocking part. I think we need to go forward with this, even if it causes issues to some snaps, as the general trend is that wayland will be used by the majority of the users.

**Comment by anonymouse64** [comment]
@zyga I just wanted to make sure that @AlanGriffiths's points were considered/understood before we went ahead with merging this, feel free to unblock this if we understand the consequences and are ready to move forward

**Comment by Erick555** [comment]
> Flatpak does not implement persistence of mount namespaces. We do and I can tell you there is a way to observe broken behaviour.

> It is sufficient for a snap to refresh and have a configure hook before a user logs in.

Hm, does it mean when for example chromium snap is refreshed before x11 session exist it will setup namespaces and leave them forever running in the background waiting for actual chromium to be launched?

**Comment by zyga** [comment]
> > Flatpak does not implement persistence of mount namespaces. We do and I can tell you there is a way to observe broken behaviour.
> 
> > It is sufficient for a snap to refresh and have a configure hook before a user logs in.
> 
> Hm, does it mean when for example chromium snap is refreshed before x11 session exist it will setup namespaces and leave them forever running in the background waiting for actual chromium to be launched?

Yes. Namespaces persist after the death of a particular process inhabiting it. Namespaces are also updated in flight, so that existing processes can observe the update on demand. This change is no different, it will propagate into existing snap namespaces even if you update snapd in the background.

**Comment by Erick555** [comment]
I guess this makes pretty hard to track files in dir like /tmp for various snaps that user may even forget they used in current session. Thx for info.

**Comment by jhenstridge** [comment]
> @zyga as this stands it breaks stuff that happens to work right now. So breaking it should be intentional.

@AlanGriffiths: Are you sure it breaks any workloads that currently work?  As it stands, snaps plugging the `x11` interface can only see the abstract namespace X11 sockets.  With the changes in this branch, they will be able to see the abstract namespace sockets, and the non-abstract sockets from one source.

If a snap has two slots connected to its `x11` plug, then I'd assume you'd have one slot side `/tmp/.X11-unix` directories mounted over the top of the other.  So while it will only see one snap's non-abstract sockets, it will be able to see both snaps' abstract sockets.

**Comment by AlanGriffiths** [comment]
> @AlanGriffiths: Are you sure it breaks any workloads that currently work? 

Yes, I give two examples above.

> If a snap has two slots connected to its `x11` plug, then I'd assume you'd have one slot side `/tmp/.X11-unix` directories mounted over the top of the other. So while it will only see one snap's non-abstract sockets, it will be able to see both snaps' abstract sockets.

If a snap has a slot connected to its `x11` plug then, with the changes in this branch, it can no longer create a non-abstract socket for (typically) an internal Xwayland instance in its private `/tmp/.X11-unix`.

**Comment by jhenstridge** [comment]
Okay.  I wonder if the best option will be to add a plug attribute to allow a snap to opt out of having `/tmp/.X11-unix` bind mounted?  That would cover the "X on X" edge cases, while still doing the right thing for the majority of snaps.  Such a snap wouldn't function if the underlying X server didn't listen on the abstract socket, but that's the case right now too.

Perhaps it could be handled automatically by turning off the mounts if the snap has any x11 slots.  I prefer the explicit option personally.

**Comment by AlanGriffiths** [comment]
> Okay. I wonder if the best option will be to add a plug attribute to allow a snap to opt out of having `/tmp/.X11-unix` bind mounted? That would cover the "X on X" edge cases, while still doing the right thing for the majority of snaps. Such a snap wouldn't function if the underlying X server didn't listen on the abstract socket, but that's the case right now too.

The most flexible option would be to mount the host `/tmp/.X11-unix` somewhere other than the snap's `/tmp/.X11-unix` (say `/tmp/.X11-unix-host`). This would allow snaps to apply the same soft-linking approach that's used for Wayland sockets avoiding the race conditions described above. I'm not sure that the added complexity is justified.

> Perhaps it could be handled automatically by turning off the mounts if the snap has any x11 slots. I prefer the explicit option personally.

There's no X11 slot for these snaps: they use `network-bind` (which doesn't need store approval).

But that does remind me that early iterations of the "X11 kiosk snap" did loopback X11 slot and plug. I don't know if there are any snaps still using this approach, or how they would be affected. Is there a way to query this on the snap store?

**Comment by jhenstridge** [comment]
> The most flexible option would be to mount the host `/tmp/.X11-unix` somewhere other than the snap's `/tmp/.X11-unix` (say `/tmp/.X11-unix-host`). This would allow snaps to apply the same soft-linking approach that's used for Wayland sockets avoiding the race conditions described above. I'm not sure that the added complexity is justified.

I don't think an approach that requires modification of every single snap with an x11 plug is workable.  So I don't think `/tmp/.X11-unix-host` is workable as a primary solution.  No existing snaps have scripts to symlink sockets on startup, and I'd like to avoid requiring them to add such scripts.

If an alternative mount point would help out for your "X on X" snaps, then perhaps that would be worth using a plug attribute for (instead of an attribute that disabled the mount).  This would give you the flexibility for the case where you need it, while letting things work by default for the vast majority of snaps.

**Comment by zyga** [comment]
I agree with @jhenstridge's opinion here. Having useful default and a way to opt out of it is IMO better.

**Comment by AlanGriffiths** [comment]
> Okay. I wonder if the best option will be to add a plug attribute to allow a snap to opt out of having `/tmp/.X11-unix` bind mounted? That would cover the "X on X" edge cases, while still doing the right thing for the majority of snaps. Such a snap wouldn't function if the underlying X server didn't listen on the abstract socket, but that's the case right now too.

The most flexible option would be to mount the host `/tmp/.X11-unix` somewhere other than the snap's `/tmp/.X11-unix` (say `/tmp/.X11-unix-host`). This would allow snaps to apply the same soft-linking approach that's used for Wayland sockets avoiding the race conditions described above. I'm not sure that the added complexity is justified.

> Perhaps it could be handled automatically by turning off the mounts if the snap has any x11 slots. I prefer the explicit option personally.

There's no X11 slot for these snaps: they use `network-bind` (which doesn't need store approval).

But that does remind me that early iterations of the "X11 kiosk snap" did loopback X11 slot and plug. I don't know if there are any snaps still using this approach, or how they would be affected. Is there a way to query this on the snap store?

> ultimately, the stuff that breaks isn't of vital importance to any usage I know of.

**Review by jhenstridge — APPROVED** [review]
I've added some spread tests on top of @zyga's work, and fixed the few small problems I ran into:

1. The `MountEntry` strucrs were had trailing slashes on the paths, which snap-update-ns rejects.
2. I converted the remaining `release.OnClassic` checks to use the `implicitSystem*Slot()` helpers.  This let me run most of the spread test on all systems: both classic and core.

The spread test uses netcat to exercise socket communication between snaps and between host and snap for unix sockets located in `/tmp/.X11-unix`.  So while it isn't using a real X server, it should tickle the right AppArmor and seccomp code paths.

**Comment by jhenstridge** [comment]
For the benefit of other reviewers, I'd consider this PR feature complete.  The idea for adding a plug attribute to modify the location the socket directory is mounted will be addressed in a follow-up PR.

**Comment by PranavBhattarai** [comment]
@jhenstridge I was wondering when exactly will Ubuntu 20.10 users can starts surfing their snap apps without having to change to X11?! I really hope it's not around 2021!

**Comment by jhenstridge** [comment]
@PranavBhattarai: there is an update for mutter currently sitting in groovy-proposed that will get snaps working on Wayland sessions without this patch:

https://launchpad.net/ubuntu/+source/mutter/3.38.1-2ubuntu1

Hopefully it'll make its way to groovy-updates soon.  I wouldn't recommend turning on proposed completely, but you could manually install the updated mutter packages if you don't want to wait.

File: interfaces/builtin/x11.go
Comment (alexmurray) [inline interfaces/builtin/x11.go:121]: Explicit denials for ICE was done in #9542 - can this comment be removed?

**Review by alexmurray — APPROVED** [review]
This looks really good - I especially like the functional socket based testing with netcat etc. Security review done.

File: interfaces/builtin/x11.go
Comment (jhenstridge) [reply interfaces/builtin/x11.go]: Good catch.  I've also removed the second TODO item about access to sockets from a non-implicit x11 slot, since the spread test demonstrates that this works on the systems where AppArmor is enabled.

**Review by mvo5 — APPROVED** [review]
Thank you!

**Pull request merged**
